hxincl.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXINCL" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxincl \- expand included HTML or XML files
.SH SYNOPSIS
.B hxincl
.RB "[\| " \-x " \|]"
.RB "[\| " \-f " \|]"
.RB "[\| " \-s
.IR name=subst " \|]"
.RB "[\| " \-s
.IR name=subst " \|]..."
.RB "[\| " \-b
.IR base " \|]"
.RI "[\| " file\-or\-URL " \|]"
.LP
.B hxincl
.B \-M
.I target
.RB "[\| " \-s
.IR name=subst " \|]"
.RB "[\| " \-s
.IR name=subst " \|]..."
.RB "[\| " \-b
.IR base " \|]"
.RB "[\| " \-G " \|]"
.RI "[\| " file\-or\-URL " \|]"
.SH DESCRIPTION
.LP
Without
.BR \-M ,
the
.B hxincl
command copies an HTML or XML file to standard output, looking for
comments with a certain structure.  Such a comment is replaced by
the file whose name is given as the attribute of the directive. For
example:
.d
\&...<!-- include "foo.html" -->...
.e
will be replaced by the content of the file \fIfoo.html\fP.
.LP
The comment is replaced by
.d
<!-- begin-include "foo.html" -->
.e
before the included text and
.d
<!-- end-include "foo.html" -->
.e
after it. These comments make it possible to run
.B hxincl
on the resulting file again to update the inclusions.
.PP
Single quotes are allowed instead of double quotes. And if the file
name contains no spaces, the quotes may also be omitted.
.PP
With
.BR \-M ,
the
.B hxincl
command outputs a line of dependencies that is suitable for inclusion
in a Makefile. The
.I target
is the target of the Makefile rule and
.B hxincl
will list after the ':' all the files that are included,
recursively. E.g., the result of
.d
hxincl -M myfile.html inputfile
.e
might be
.d
myfile.html: foo.html bar.html
.e
.SH OPTIONS
The following options are supported:
.TP 10
.B \-x
Use XML conventions: empty elements are written with a slash at the
end: <IMG\ />.
.TP
.BI \-b " base"
Sets the base URL for resolving relative URLs. By default the file
given as argument is the base URL.
.TP
.B \-f
Removes the comments after including the files. This means
.B hxincl
cannot be run on the resulting file later to update the inclusions.
(Mnemonic:
.BR f inal
or
.BR f rozen.)
.TP
.BI \-s " name=substitution"
Include a different file than the one mentioned in the directive. If
the comment is
.d
<!-- include "name" -->
.e
the file
.I substitution
is included instead. And if the file name in the comment includes a
variable called
.I name
delimited by %, e.g.,
.d
<!-- include "xxx-%name%" -->
.e
then
.RI % name %
is replaced by
.I substitution
and thus the file
.RI xxx\- substitution
is included. The option
.B \-s
may occur multiple times. %-delimited variables are expanded
recursively, i.e., if the substitution text contains a variable, that
variable is expanded, too. E.g., if the two options
.B \-s name=%p1%.rrr
and
.B \-s p1=subst
are given, then the "xxx-%name%" will expand to "xxx-subst.rrr".
.TP
.BI \-M " target"
Instead of outputting the input file with all inclusions expanded,
output just the list of all files that the input includes,
recursively, in the form of a rule that is suitable for a
Makefile. The
.I target
is printed as the target of that rule.
.TP
.B \-G
Suppress error messages if a file to include cannot be found. (Only
with
.BR \-M .)
.SH OPERANDS
The following operand is supported:
.TP 10
.I file\-or\-URL
The name of an HTML or XML file or the URL of one. If absent, standard
input is read instead.
.SH "EXIT STATUS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred in the parsing of one of the HTML or XML files.
.SH ENVIRONMENT
To use a proxy to retrieve remote files, set the environment variables
.B http_proxy
or
.BR ftp_proxy "."
E.g.,
.B http_proxy="http://localhost:8080/"
.SH BUGS
.LP
Assumes UTF-8 as input. Doesn't expand character entities. Instead
pipe the input through
.BR hxunent (1)
and
.BR asc2xml (1)
to convert it to UTF-8.
.LP
Remote files (specified with a URL) are currently only supported for
HTTP. Password-protected files or files that depend on HTTP "cookies"
are not handled. (You can use tools such as
.BR curl (1)
or
.BR wget (1)
to retrieve such files.)
.SH "SEE ALSO"
.BR asc2xml (1),
.BR hxnormalize (1),
.BR hxnum (1),
.BR hxprune (1),
.BR hxtoc (1),
.BR hxunent (1),
.BR xml2asc (1),
.BR UTF-8 " (RFC 2279)"


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxincl.c
/*
 * incl - expand included files
 *
 * Searches for <!--include "file"--> and expands the referenced file
 * in place. File may be a URL. Works recursively. Other accepted
 * syntaxes:
 *
 * <!--include "file"-->
 * <!--include 'file'-->
 * <!--include file-->
 * <!--begin-include "file"-->...<!--end-include-->
 * <!--begin-include 'file'-->...<!--end-include-->
 * <!--begin-include file-->...<!--end-include-->
 *
 * If there are no quotes, the file name may not include whitespace.
 *
 * Copyright © 1994-2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos
 * Created: 2 Dec 1998
 * Version: $Id: hxincl.c,v 1.14 2017/11/24 09:50:25 bbos Exp $
 *
 **/
#include "config.h"
#include <stdio.h>
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include <ctype.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
# ifndef HAVE_STRSTR
#  include "strstr.e"
# endif
#endif
#include <stdlib.h>
#include <assert.h>
#include <err.h>
#include <stdbool.h>
#include "export.h"
#include "types.e"
#include "errexit.e"
#include "html.e"
#include "scan.e"
#include "dict.e"
#include "openurl.e"
#include "heap.e"
#include "url.e"

#define INCLUDE "include"
#define BEGIN "begin-include"
#define END "end-include"

typedef struct _stack {
  bool skipping;
  struct _stack *next;
} *stack;

typedef enum {KNone, KIncl, KBegin, KEnd} Key;

static bool do_xml = false;
static bool final = false;	/* Final means to remove the comments */
static bool has_error = false;
static stack skipping = NULL;
static Dictionary substitutions = NULL;
static string target = NULL;
static bool warn_missing = true;


/* push -- push a skipping state on the stack */
static void push(stack *skipping, bool s)
{
  stack h;

  new(h);
  h->next = *skipping;
  h->skipping = s;
  *skipping = h;
}

/* pop -- pop a skipping state off the stack */
static void pop(stack *skipping)
{
  stack h;

  assert(*skipping);
  h = *skipping;
  *skipping = (*skipping)->next;
  dispose(h);
}

/* top -- return value of top of skipping stack */
static bool top(stack skipping)
{
  assert(skipping);
  return skipping->skipping;
}

/* word_to_key -- check whether word s is one of the recognized keywords */
static Key word_to_key(const string s, int len)
{
  if (len == sizeof(END) - 1 && strncmp(s, END, len)== 0) return KEnd;
  if (len == sizeof(INCLUDE) - 1 && strncmp(s, INCLUDE, len)== 0) return KIncl;
  if (len == sizeof(BEGIN) - 1 && strncmp(s, BEGIN, len)== 0) return KBegin;
  return KNone;
}

/* add_substitution -- add a file name substitution to the dictionary */
static void add_substitution(const string assignment)
{
  string s;

  /* To do: handle file names containing '='. Add escapes? */
  if (!substitutions) substitutions = dict_create(10);
  if (!substitutions) errexit("Out of memory.\n");
  if (!(s = strchr(assignment, '='))) errexit("No '=' found in option -s\n");
  *s = '\0';
  if (!dict_add(substitutions, assignment, s + 1)) errexit("Out of memory?\n");
  *s = '=';
}

/* expand_vars -- look for and expand %variables% in s */
static string expand_vars(const conststring s)
{
  string h, k, sub2, var, result = NULL;
  conststring sub;

  k = strdup(s);
  if (!k) err(1, NULL);

  /* To do: avoid infinite loops */
  for (h = k; h;) {
    /* Append the text leading up to next '%' */
    result = strapp(&result, strsep(&h, "%"), NULL);
    /* If there is any text left, get the text until the next '%' into var */
    if ((var = strsep(&h, "%"))) {
      if (!h) {			/* No matching '%' found */
	result = strapp(&result, "%", var, NULL);
      } else if (*var == '\0') { /* Treat "%%" as a single "%" */
	result = strapp(&result, "%", NULL);
      } else if (substitutions && (sub = dict_find(substitutions, var))) {
	sub2 = expand_vars(sub);
	result = strapp(&result, sub2, NULL);
	free(sub2);
      } else {			/* Undefined %var% */
	result = strapp(&result, "%", var, "%", NULL);
      }
    }
  }
  free(k);
  return result;
}

/* --------------- implements interface api.h -------------------------- */

/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_error = true;
}

/* start -- called before the first event is reported */
void* start(void)
{
  push(&skipping, false);			/* Start by not skipping */
  return NULL;
}
  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  assert(clientdata == NULL);
  assert(top(skipping) == false);
  /* skip */
}

/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  /* A push() occurs at <!--begin-include...--> and at include_file() */
  /* A pop() occurs at <!--end-include...--> and at ENDINCL */
  int i, j, status;
  conststring s, url;
  FILE *f;
  Key key;

  i = strspn(commenttext, " \t\n\r\f");		/* Skip whitespace */
  j = strcspn(commenttext + i, " \t\n\r\f");	/* First word */
  key = word_to_key(commenttext + i, j);
  
  if (key == KEnd) {				/* <!--end-include...--> */

    /* Don't print anything, just pop a level */
    pop(&skipping);

  } else if (top(skipping)) {			/* Are we already skipping? */

    /* Don't print anything; push a level if this is a begin-include */
    if (key == KBegin) push(&skipping, true);

  } else if (key == KNone) {			/* Unrecognized comment? */

    /* Print the comment verbatim */
    if (!target) printf("<!--%s-->", commenttext);

  } else {					/* include or begin-include */

    /* Push a level if this is a begin-include */
    if (key == KBegin) push(&skipping, true);

    /* Find start of file name */
    i += j;
    i += strspn(commenttext + i, " \t\n\r\f");	/* Skip whitespace */

    /* Accept either "...", '...', or any string without spaces */
    if (commenttext[i] == '"') {
      j = strcspn(commenttext + i + 1, "\"");
      url = newnstring(commenttext + i + 1, j);
    } else if (commenttext[i] == '\'') {
      j = strcspn(commenttext + (++i), "'");
      url = newnstring(commenttext + i + 1, j);
    } else {
      j = strcspn(commenttext + i, " \t\n\r\f");
      url = newnstring(commenttext + i, j);
    }

    /* If we have a substitution for it, use that instead */
    if (substitutions && (s = dict_find(substitutions, url))) url = s;

    /* Expand any %variables% in the url */
    url = expand_vars(url);

    /* Get the file and recursively parse it */
    assert(get_yyin_name());
    s = URL_s_absolutize(get_yyin_name(), url);
    if (target) printf(" \\\n %s", s); /* To do: escape spaces in s */
    if (!(f = fopenurl(s, "r", &status))) {
      if (!target || warn_missing) perror(url);
    } else if (status != 200) {
      if (!target || warn_missing)
	fprintf(stderr, "%s : %s\n", url, http_strerror(status));
    } else {
      if (!final && !target) printf("<!--%s %s-->", BEGIN, commenttext + i);
      push(&skipping, false);
      include_file(f, s);
    }
    dispose(url);
    dispose(s);
  }

  free(commenttext);
}

/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  if (top(skipping) == false && !target) printf("%s", text);
  free(text);
}

/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi, string fpi, string url)
{
  if (!target) {
    printf("<!DOCTYPE %s", gi);
    if (fpi) printf(" PUBLIC \"%s\"", fpi);
    if (url) printf(" %s\"%s\"", fpi ? "" : "SYSTEM ", url);
    printf(">");
  }
  free(gi);
  free(fpi);
  free(url);
}

/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  if (top(skipping) == false && !target) printf("<?%s>", pi_text);
  free(pi_text);
}

/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  pairlist p;

  if (top(skipping) == false && !target) {
    printf("<%s", name);
    for (p = attribs; p; p = p->next) {
      if (p->value != NULL) printf(" %s=\"%s\"", p->name, p->value);
      else if (do_xml) printf(" %s=\"%s\"", p->name, p->name);
      else printf(" %s", p->name);
    }
    printf(">");
  }
  free(name);
  pairlist_delete(attribs);
}

/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  pairlist p;

  if (top(skipping) == false && !target) {
    printf("<%s", name);
    for (p = attribs; p; p = p->next) {
      if (p->value != NULL) printf(" %s=\"%s\"", p->name, p->value);
      else if (do_xml) printf(" %s=\"%s\"", p->name, p->name);
      else printf(" %s", p->name);
    }
    printf(do_xml ? " />" : ">");
  }
  free(name);
  pairlist_delete(attribs);
}

/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  if (top(skipping) == false && !target) printf("</%s>", name);
  free(name);
}

/* handle_endincl -- called after the end of an included file is reached */
void handle_endincl(void *clientdata)
{
  pop(&skipping);

  /* Mark the end of the inclusion */
  if (!final && !target) printf("<!--%s-->", END);
}

/* --------------------------------------------------------------------- */

/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "Usage: %s [-v] [-x] [-b base] [-s name=subst ...] [-M target [-G]] [file-or-url]\n",
	  prog);
  exit(2);
}

int main(int argc, char *argv[])
{
  int c, status = 200;
  string base = NULL;

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);
  set_endincl_handler(handle_endincl);

  /* Parse command line arguments */
  while ((c = getopt(argc, argv, ":xb:fs:M:Gv")) != -1)
    switch (c) {
    case 'x': do_xml = true; break;
    case 'b': base = optarg; break;
    case 'f': final = true; break;
    case 's': add_substitution(optarg); break;
    case 'M': target = optarg; break;
    case 'G': warn_missing = false; break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    default: usage(argv[0]);
    }

  if (optind == argc) {
    set_yyin(stdin, base ? base : "stdin");
  } else if (optind == argc - 1) {
    set_yyin(fopenurl(argv[optind], "r", &status), base ? base : argv[optind]);
  } else {
    usage(argv[0]);
  }

  if (yyin == NULL) {perror(argv[optind]); exit(1);}
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  if (target) printf("%s:", target);
  if (yyparse() != 0) exit(3);
  if (target) printf("\n");

  return has_error ? 1 : 0;
}
